[Autoscaling] Add end user documentation (#4337)
elastic/cloud-on-k8s: docs/orchestrating-elastic-stack-applications/elasticsearch/autoscaling.asciidoc

:parent_page_id: elasticsearch-specification
:page_id: autoscaling
ifdef::env-github[]
****
link:https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-{parent_page_id}.html#k8s-{page_id}[View this document on the Elastic website]
****
endif::[]
[id="{p}-{page_id}"]
= Elasticsearch autoscaling

experimental[]

> ECK can leverage the autoscaling API introduced in Elasticsearch 7.11 to adjust automatically the number of Pods and the allocated resources in a tier. Currently, autoscaling is supported for Elasticsearch data tiers and machine learning nodes.

Reviewer: I wonder if we should provide a little more explanation about what is a tier because it's pretty new. Maybe just add this link https://www.elastic.co/guide/en/elasticsearch/reference/current/data-tiers.html?

> ECK can leverage the autoscaling API introduced in Elasticsearch 7.11 to adjust automatically the number of Pods and the allocated resources in a tier. Currently, autoscaling is supported for Elasticsearch link:https://www.elastic.co/guide/en/elasticsearch/reference/current/data-tiers.html[data tiers] and machine learning nodes.

Reviewer 2: Should we link to https://www.elastic.co/guide/en/elasticsearch/reference/master/autoscaling-apis.html here?

[float]
[id="{p}-enable"]
== Enable autoscaling

> To enable autoscaling on an Elasticsearch cluster, you need to define one or more autoscaling policies. Each autoscaling policy applies to one or mode `NodeSets` which share the same set of roles specified in the `node.roles` setting in the Elasticsearch configuration.

Reviewer: typo:
```suggestion
To enable autoscaling on an Elasticsearch cluster, you need to define one or more autoscaling policies. Each autoscaling policy applies to one or more `NodeSets` which share the same set of roles specified in the `node.roles` setting in the Elasticsearch configuration.
```

[float]
[id="{p}-{page_id}-policies"]
=== Define autoscaling policies

Autoscaling policies can be defined in the `elasticsearch.alpha.elastic.co/autoscaling-spec` annotation of the Elasticsearch resource. Each autoscaling policy must have the following fields:

* `name` is a unique name used to identify the autoscaling policy.

> * `roles` contains a set of node roles, unique across all the autoscaling policies, used to identify the `NodeSets` to which this policy applies. At least one `NodeSet` with the exact same set of roles must exist in the Elasticsearch resource specification.

Reviewer: Nit: we haven't used backquotes around NodeSet in other [doc pages](https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-orchestration.html#k8s-nodesets), so let's stay consistent?

* `resources` helps define the minimum and maximum compute resources usage:
** `nodeCount` defines the minimum and maximum nodes allowed in the tier.
** `cpu` and `memory` enforce minimum and maximum compute resources usage for the Elasticsearch container.
** `storage` enforces minimum and maximum storage request per PersistentVolumeClaim.

[source,json]
----
{
    "policies": [{

>       "name": "di",

Reviewer 2: Should we use a more speaking name here? Eg.`data-ingest`

      "roles": ["data", "ingest" , "transform"],
      "resources": {
          "nodeCount": { "min": 3, "max": 8 },
          "cpu": { "min": 2, "max": 8 },
          "memory": { "min": "2Gi", "max": "16Gi" },
          "storage": { "min": "64Gi", "max": "512Gi" }
      }
    },
    {
      "name": "ml",
      "roles": ["ml"],
      "resources": {
          "nodeCount": { "min": 1, "max": 9 },
          "cpu": { "min": 1, "max": 4 },
          "memory": { "min": "2Gi", "max": "8Gi" }
      }
    }]
}
----

> WARNING: A node role should not be referenced in more than one autoscaling policy.

Reviewer 2: Warning implies that something "bad" happens if users do not heed the warning. But in this case iiuc we are 
1. validating that there is not overlap in policies
2. even if the webhook is disabled we recheck in the reconciliation loop and issue events

So maybe a `NOTE` would be more appropriate?

Author: There is no validation, it might be a temporary situation in the case of a transition between different tier models for example (e.g. from co-located roles to dedicated nodes per role).
For now, when detected, [this situation is reported in the logs and in the autoscaling status](https://github.com/elastic/cloud-on-k8s/pull/4265).

For example, you can remove the memory limit and set a CPU limit to twice the value of the request, as follows:

[source,json]
----
{
    "policies": [{
        "name": "di-hot",
        "roles": ["data_hot", "ingest", "transform"],
        "resources": {
            "nodeCount": {
                "min": 2,
                "max": 5
            },
            "cpu": {
                "min": 1,
                "max": 2,
                "requestsToLimitsRatio": 2
            },
            "memory": {
                "min": "2Gi",
                "max": "6Gi",
                "requestsToLimitsRatio": 0
            }
        }
    }]
}
----

You can find link:{eck_github}/blob/{eck_release_branch}/config/recipes/autoscaling/elasticsearch.yaml[a complete example in the ECK GitHub repository] which will also show you how to fine-tune the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/autoscaling-deciders.html[autoscaling deciders].

[float]
[id="{p}-{page_id}-polling-interval"]
=== Change the polling interval

> The Elasticsearch autoscaling capacity endpoint is polled every minute. This interval duration can be controlled using the `pollingPeriod` field in the autoscaling specification:

Reviewer: It might be good to be more specific here, although it may seem obvious to us.
```suggestion
The Elasticsearch autoscaling capacity endpoint is polled every minute by the operator. This interval duration can be controlled using the `pollingPeriod` field in the autoscaling specification:
```

[source,json]
----
{
	"policies": [
		{
			"name": "di-hot",
			"nodeSets": [{
				"name": "di-hot",
				"nodeCount": 5
			}],
			"resources": {
				"limits": {
					"cpu": "2",
					"memory": "6Gi"
				},
				"requests": {
					"cpu": "2",
					"memory": "6Gi",
					"storage": "6Gi"
				}
			},
			"state": [{
				"type": "HorizontalScalingLimitReached",
				"messages": [
					"Can't provide total required storage 32588740338, max number of nodes is 5, requires 6 nodes"
				]
			}],
			"lastModificationTime": "2021-03-09T17:01:25Z"
		}
	]
}
----

Important events are also reported through Kubernetes events, for example when the maximum autoscaling size limit is reached:

[source,sh]
----
kubectl get events
----

[source,sh]
----
40m  Warning  HorizontalScalingLimitReached  elasticsearch/sample   Can't provide total required storage 32588740338, max number of nodes is 5, requires 6 nodes

> ----

Reviewer 2: Should this be one code block?

[float]
[id="{p}-disable"]
== Disable autoscaling

You can disable autoscaling at any time by removing the `elasticsearch.alpha.elastic.co/autoscaling-spec` annotation from the Elasticsearch resource metadata.

For machine learning the following settings are not automatically reset:

- `xpack.ml.max_ml_node_size`
- `xpack.ml.max_lazy_ml_nodes`
- `xpack.ml.use_auto_machine_memory_percent`

> You should set those settings manually according to your deployment.

Reviewer 2: ```suggestion
You should adjust those settings manually to match the size of your deployment when you disable autoscaling.
```
